qgym/benchmarks/metrics/initial_mapping_metrics.py

```diff
@@ -0,0 +1,146 @@
+"""Metrics to assess the performance of InitialMappers.
+
+Functions named as ``*_score`` return a scalar value to maximize: the higher
+the better.
+
+Function named as ``*_error`` or ``*_loss`` return a scalar value to minimize:
+the lower the better.
+"""
+
+from __future__ import annotations
+
+from collections import deque
+from collections.abc import Iterable
+from copy import deepcopy
+from typing import Protocol, cast, runtime_checkable
+
+import networkx as nx
+import numpy as np
+from numpy.typing import ArrayLike, NDArray
+from qiskit import QuantumCircuit
+from qiskit.dagcircuit import DAGCircuit
+
+from qgym.generators.graph import BasicGraphGenerator, GraphGenerator
+from qgym.utils.input_validation import check_string
+
+
+@runtime_checkable
+class InitialMappingMetric(Protocol):
+    def compute(self, interaction_graph: nx.Graph, mapping: ArrayLike) -> float: ...
+
+
+@runtime_checkable
+class Mapper(Protocol):
+    def compute_mapping(
+        self, circuit: QuantumCircuit | DAGCircuit
+    ) -> NDArray[np.int_]: ...
+
+
+class DistanceRatioLoss(InitialMappingMetric):
+
+    def __init__(self, connection_graph: nx.Graph) -> None:
+        """Init of the :class:`InitialMappingSolutionQuality` class.
+
+        Args:
+            connection_graph: :class:`networkx.Graph` representation of the QPU
+                topology. Each node represents a physical qubit and each edge represents
+                a connection in the QPU topology.
+
+        """
+        self.connection_graph = connection_graph
+        n_qubits = len(self.connection_graph)
+        self.distance_matrix = np.zeros((n_qubits, n_qubits), dtype=np.int_)
+        for node_u, distances in nx.all_pairs_shortest_path_length(connection_graph):
+            for node_v in connection_graph:
+                self.distance_matrix[node_u, node_v] = distances[node_v]
+
+    def compute(self, interaction_graph: nx.Graph, mapping: ArrayLike) -> float:
+        if interaction_graph.number_of_edges() == 0:
+            return 1.0
+        mapping = np.asarray(mapping, dtype=np.int_)
+        distance_loss = 0
+        for edge in interaction_graph.edges():
+            mapped_edge = (mapping[edge[0]], mapping[edge[1]])
+            distance_loss += self.distance_matrix[mapped_edge]
+
+        return float(distance_loss / interaction_graph.number_of_edges())
+
+
+class AgentPerformance:
+
+    def __init__(
+        self,
+    ) -> None:
+        """Init of the :class:`AgentPerformance` class.
+
+        Args:
+        """
+
+
+class InitialMappingBenchmarker:
+
+    def __init__(
+        self,
+        generator: GraphGenerator | None = None,
+        *,
+        metrics: Iterable[InitialMappingMetric] | InitialMappingMetric,
```

Comment:
Could this always be an Iterable? We would force the caller to put a single metric into a list, but hey, that's our API. It would simplify the code internally.